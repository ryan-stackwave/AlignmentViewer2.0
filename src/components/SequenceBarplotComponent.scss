/** SequenceBarplotComponent.scss */

.barplot.hovered .barTooltip {
  display: block;
}
.barplot .barTooltip {
  position: absolute;
  font-size: 10px;
  background: white;
  display: none;
}

.barplot.hovered .legend {
  display: block;
}
.barplot .legend {
  display: none;
  background-color: rgba(255, 255, 255, 0.9);
  border: solid 1px black;
  position: absolute;
  padding: 3px;
  top: 0;
  right: 0;
  z-index: 1000;
}

.barplot svg rect {
  cursor: pointer;
}

.barplot svg .barplot-entropy rect.selected,
.barplot svg .barplot-gaps rect.selected,
.barplot svg .barplot-category3 rect.selected {
  fill: red;
}
.barplot .legend .barplot-entropy,
.barplot svg .barplot-entropy rect {
  fill: #000000;
  color: #000000;
}
.barplot .legend .barplot-gaps,
.barplot svg .barplot-gaps rect {
  fill: #b7b7b7;
  color: #b7b7b7;
}
.barplot .legend .barplot-category3,
.barplot svg .barplot-category3 rect {
  fill: green;
  color: green;
}
