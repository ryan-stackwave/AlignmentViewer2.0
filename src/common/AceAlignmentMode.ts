/**
 * AceAlignmentMode.js
 * This file defines an alignment highlight mode form ace editor documents.
 */

import { default as ace } from "ace-builds";
import { Alignment } from "./Alignment";
import { getLetterClassNames } from "./MolecularStyles";

interface IAceType {
  [key: string]: any;
}

export function defineNewAlignmentMode(mode: string, alignment: Alignment) {
  const modeSplit = mode.split("/");
  if (
    modeSplit.length !== 3 ||
    modeSplit[0] !== "ace" ||
    modeSplit[1] !== "mode"
  ) {
    throw Error("ace mode name must be of the form '/ace/mode/MODENAME'");
  }
  const modeName = mode.split("/")[2];

  (ace as IAceType).define(
    "ace/mode/" + modeName + "_highlight_rules",
    function (require: Function, exports: any, module: any) {
      var oop = require("../lib/oop");
      var TextHighlightRules = require("./text_highlight_rules")
        .TextHighlightRules;

      /**
       * Rules parse individual characters in the editor and assign classes
       * to each character. Note that characters will be grouped together
       * if they match the same rule and are side by side.
       */
      var AlignmentHighlightRules = function (this: any) {
        //We define an alignment-specific set of rules that will
        //add classes including whether a character matches the
        //consensus sequence, matches the query etc.
        let rules: {
          [key: string]: {
            token: (letter: string) => void;
            regex: RegExp;
            next: string;
          }[];
        } = {};

        //note: this is by far the thing that kills performance the most
        //      I'm not sure whether it is due to how it parses (state machine)
        //      slowing things down, or whether it is an issue with attaching
        //      ~4 classes to every single residue -- either way, removing parsing
        //      all together makes it much more snappy and also decreases the
        //      memory footprint drastically.
        const re = /./;
        const statePrefix = "p";
        for (let i = 0; i < alignment.getMaxSequenceLength(); i++) {
          const stateName = i === 0 ? "start" : statePrefix + i;
          const nextStateName =
            i === alignment.getMaxSequenceLength() - 1
              ? "start"
              : statePrefix + (i + 1);

          rules[stateName] = [
            {
              token: function (letter: string) {
                const toreturn = getLetterClassNames(
                  letter,
                  letter === alignment.getConsensus()[i].letter,
                  letter === alignment.getTargetSequence().sequence[i],
                  true
                );
                return toreturn;
              },
              regex: re,
              next: nextStateName,
            },
          ];
        }
        this.$rules = rules;
      };

      oop.inherits(AlignmentHighlightRules, TextHighlightRules);

      exports.AlignmentHighlightRules = AlignmentHighlightRules;
    }
  );

  (ace as IAceType).define("ace/mode/" + modeName, function (
    require: Function,
    exports: any,
    module: any
  ) {
    var oop = require("../lib/oop");
    var TextMode = require("../mode/text").Mode;
    var Tokenizer = require("../tokenizer").Tokenizer;
    var AlignmentHighlightRules = require("ace/mode/" +
      modeName +
      "_highlight_rules").AlignmentHighlightRules;

    const Mode = function (this: any) {
      this.$tokenizer = new Tokenizer(new AlignmentHighlightRules().getRules());
    };

    oop.inherits(Mode, TextMode);

    (function () {
      // Extra logic goes here. (see below)
    }.call(Mode.prototype));

    exports.Mode = Mode;
  });
}
