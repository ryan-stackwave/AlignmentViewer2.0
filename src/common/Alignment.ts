import { generateUUIDv4 } from "./Utils";
import { defineNewAlignmentMode } from "./AceAlignmentMode";
import { Nucleotide } from "./Residues";
import { SequenceSorter } from "./AlignmentSorter";
import {
  AlignmentTypes,
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
} from "./MolecularStyles";

export interface ISequence {
  id: string;
  sequence: string;
}

/**
 * Alignment
 * This class represents a multiple sequence alignment. During initialization,
 * positional statistics are generated about the alignment and a consensus
 * sequence is determined.
 *
 * The class should generally be instantiated from subclasses (e.g., FastaAlignemtn)
 * using static methods, e.g.,
 *     FastaAlignment.fromFileContents
 *     StockholmAlignment.fromFileContents
 */
export class Alignment {
  private uuid: string;
  private aceEditorMode: string | undefined;
  private name: string;
  private predictedNT: boolean;
  private sequences: Map<SequenceSorter, ISequence[]>;
  private querySequence: ISequence;
  private positionalLetterCounts = new Map<
    number,
    { [letter: string]: number }
  >();
  private globalAlphaLetterCounts: { [letter: string]: number } = {};
  private allUpperAlphaLettersInAlignmentSorted: string[];
  private consensus: {
    sequence: ISequence;
    statistics: {
      letter: string;
      position: number;
      occurrences: number;
    }[];
  };

  /**
   * Normalize all values in an object that contains all counts such that
   * the sum of those counts equals 1.
   * @param letterCounts an object that contains the same keys, but whose
   *                     values now sum to 1.
   * @param validLetters if provided, returned object will only contain letters
   *                     that are in this array.
   */
  private static normalizeLetterCounts(
    letterCounts: { [letter: string]: number },
    validLetters?: string[]
  ) {
    const totalCounts = Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc += letterCounts[letter];
      }
      return acc;
    }, 0);

    return Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc[letter] = letterCounts[letter] / totalCounts;
      }
      return acc;
    }, {} as { [letter: string]: number });
  }

  /**
   * Create a new Alignment object. The default query sequence will
   * be set to the first sequence in sequences. This can be changed
   * by calling setQuerySequence() after creation
   * @param sequences
   */
  public constructor(name: string, sequencesAsInput: ISequence[]) {
    this.uuid = generateUUIDv4();
    this.name = name;
    this.sequences = new Map<SequenceSorter, ISequence[]>();
    this.sequences.set(SequenceSorter.INPUT, sequencesAsInput);
    this.querySequence = sequencesAsInput[0];
    this.predictedNT = true;

    //
    //generate statistics
    //
    const start = new Date();
    const allLettersInAlignment: { [key: string]: number } = {}; //all letters in the alignment

    // aggregate stats for each position and globally including
    // the number of time each amino acid occurs
    // ** this loop takes the bulk of the initialization time
    for (
      let sequenceIdx = 0;
      sequenceIdx < sequencesAsInput.length;
      sequenceIdx++
    ) {
      const seq = sequencesAsInput[sequenceIdx].sequence;
      for (let positionIdx = 0; positionIdx < seq.length; positionIdx++) {
        const letter = seq[positionIdx];
        const letterIsAlpha = letter.match(/[a-z]/i) ? true : false;
        if (
          letterIsAlpha &&
          this.predictedNT &&
          Nucleotide.fromSingleLetterCode(letter) === Nucleotide.UNKNOWN
        ) {
          this.predictedNT = false;
        }

        const position = positionIdx; // zero based
        allLettersInAlignment[letter] = 1;

        let letterCounts = this.positionalLetterCounts.get(position);
        if (!letterCounts) {
          letterCounts = {};
          this.positionalLetterCounts.set(position, letterCounts);
        }
        if (letter in letterCounts === false) {
          letterCounts[letter] = 0;
        }
        letterCounts[letter] += 1;

        if (letterIsAlpha) {
          if (letter in this.globalAlphaLetterCounts === false) {
            this.globalAlphaLetterCounts[letter] = 0;
          }
          this.globalAlphaLetterCounts[letter] += 1;
        }
      }
    }

    //get a list of all the upper letters found in the alignment
    this.allUpperAlphaLettersInAlignmentSorted = Object.keys(
      allLettersInAlignment
    )
      .sort()
      .reduce((arr: string[], value: string) => {
        if (value.match(/[A-Z]/)) {
          // only keep letters
          arr.push(value);
        }
        return arr;
      }, []);

    //determine consensus sequence
    const consensusStats = Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        const topLetter = Object.keys(letterCounts).reduce((a, b) => {
          const aIsLetter = a.match(/[a-z]/i);
          const bIsLetter = b.match(/[a-z]/i);
          if (aIsLetter && !bIsLetter) {
            return a;
          }
          if (bIsLetter && !aIsLetter) {
            return b;
          }
          return letterCounts[a] > letterCounts[b] ? a : b;
        });
        return {
          position: position,
          letter: topLetter,
          occurrences: letterCounts[topLetter],
        };
      }
    );
    this.consensus = {
      sequence: {
        id: "consensus",
        sequence: consensusStats.map((s) => s.letter).join(""),
      },
      statistics: consensusStats,
    };

    console.log(
      "done parsing alignment. took " +
        (new Date().getTime() - start.getTime()) +
        "ms"
    );
  }

  /**
   * Is this alignment predicted to be nucleotides?
   * @returns true if it is predicted to be nucleotide sequences.
   */
  getPredictedType(): AlignmentTypes {
    return this.predictedNT
      ? AlignmentTypes.NUCLEOTIDE
      : AlignmentTypes.AMINOACID;
  }

  /**
   * Returns a default nucleotide or amino acid style for this alignment,
   * predicted from the sequence itself.
   */
  getDefaultStyle(): AminoAcidAlignmentStyle | NucleotideAlignmentStyle {
    return this.predictedNT
      ? new NucleotideAlignmentStyle()
      : new AminoAcidAlignmentStyle();
  }

  /**
   * Get a unique uuid for this alignment.
   * @returns a uuid string for this alignment
   */
  getUUID() {
    return this.uuid;
  }

  /**
   * Get the name of this alignment
   * @returns a name for the alignment - usually the filename
   */
  getName() {
    return this.name;
  }

  /**
   * Get an ace editor mode for this alignemnt. If one doesn't exist
   * then it will be defined.
   * @returns the ace editor mode for this alignment
   */
  getAceEditorMode() {
    if (!this.aceEditorMode) {
      this.aceEditorMode = "ace/mode/" + this.uuid;
      defineNewAlignmentMode(this.aceEditorMode, this);
    }
    return this.aceEditorMode;
  }

  /**
   * Get the consensus sequence
   * @returns a position ordered array of objects that contain each letter, its number of
   *          occurrences, and the actual position number
   */
  getConsensus() {
    return this.consensus;
  }

  /**
   * Get the number of gaps at each position
   * @returns an array of gap counts, ordered by position
   */
  getGapsPerColumn(): number[] {
    return Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return !isNaN(letterCounts["-"]) ? letterCounts["-"] : 0;
      }
    );
  }

  /**
   * Get the number of gaps at a specific position
   * @returns the gap counts at the specified position
   */
  getGapCountAtColumn(position: number): number {
    const letterCounts = this.positionalLetterCounts.get(position);
    if (!letterCounts || !letterCounts["-"]) {
      return 0;
    }
    return letterCounts["-"];
  }

  /**
   * Get the total number of times each letter appears in the alignment.
   * @param normalize if true, will normalize the returned array such that
   *                  it sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   */
  getGlobalAlphaLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize) {
      return Alignment.normalizeLetterCounts(
        this.globalAlphaLetterCounts,
        validLetters
      );
    }
    return !validLetters
      ? this.globalAlphaLetterCounts
      : Object.keys(this.globalAlphaLetterCounts).reduce((acc, letter) => {
          //don't normalize, but remove invalid letters
          if (validLetters.includes(letter)) {
            acc[letter] = this.globalAlphaLetterCounts[letter];
          }
          return acc;
        }, {} as { [letter: string]: number });
  }

  /**
   * Get the length of the longest sequence
   * @returns the length of the largest sequence in this alignment
   */
  getMaxSequenceLength(): number {
    return this.getSequences().reduce((acc, seq) => {
      if (seq.sequence.length > acc) {
        acc = seq.sequence.length;
      }
      return acc;
    }, -1);
  }

  /**
   * Get the frequencies of all letters in this alignment for each position.
   * @param normalize if true, will normalize the returned letter counts for each
   *                  position such that each sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   * @returns a Map whose keys are position (ordered) and values are a dictionary
   *          with [key = letter (e.g., amino acid code)] and [value = # occurrences].
   * TODO: Test parameter use cases.
   */
  getPositionalLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize || validLetters) {
      return Array.from(this.positionalLetterCounts).reduce(
        (acc, [position, letterCounts]) => {
          acc.set(
            position,
            normalize
              ? Alignment.normalizeLetterCounts(letterCounts, validLetters) // normalize + remove invalid letters
              : Object.keys(letterCounts).reduce((acc2, letter) => {
                  //don't normalize, but remove invalid letters
                  if (validLetters!.includes(letter)) {
                    acc2[letter] = letterCounts[letter];
                  }
                  return acc2;
                }, {} as { [letter: string]: number })
          );
          return acc;
        },
        new Map<Number, { [letter: string]: number }>()
      );
    }
    return this.positionalLetterCounts;
  }

  /**
   * Get all the sequences in this alignment - ordered as they were input
   * @param sortBy specifies the order of the returned sequences. If it isn't
   *               provided, returns the sequences as they were input to create
   *               the Alignment object, e.g., usually fasta order.
   * @param forceReSort if provided, sequences are forced to be re-sorted (rather
   *                    than possibly cached)
   * @returns all sequences in this alignment
   */
  getSequences(sortBy?: SequenceSorter, forceReSort?: boolean): ISequence[] {
    sortBy = sortBy ? sortBy : SequenceSorter.INPUT;
    forceReSort = forceReSort ? forceReSort : false;
    if (
      sortBy !== SequenceSorter.INPUT &&
      (!this.sequences.has(sortBy) || forceReSort)
    ) {
      //cache not yet populated for this sort option
      this.sequences.set(
        sortBy,
        sortBy.sortFn(this.sequences.get(SequenceSorter.INPUT)!, this)
      );
    }
    return this.sequences.get(sortBy)!;
  }

  /**
   * Get a sorted list of all of the upper case letters in the alignment.
   * @returns a list of all letters in the alignment sorted alphabetically.
   */
  getAllUpperAlphaLettersInAlignmentSorted(): string[] {
    return this.allUpperAlphaLettersInAlignmentSorted;
  }

  /**
   * Get all the first sequence in the alignment, aka, target / query / first
   * @returns all sequences in this alignment
   */
  getQuerySequence(): ISequence {
    return this.querySequence;
  }

  /**
   * Set the query sequence
   * @param newQuery The new query sequence. Must already exist in the alignment.
   * @throws an error if the query sequence does not yet exist in the alignment
   *
   * IMPORTANT: Requires re-sorting
   */
  /*setQuerySequence(newQuery: ISequence) {
    const queryInAlignment = this.getSequences().find((seq) => {
      if (seq === newQuery) {
        return true;
      }
      return false;
    });
    if (queryInAlignment !== undefined) {
      this.querySequence = newQuery;

      //the resort could be really slow ... need to rethink.
      SequenceSorter.list.forEach((sortOption) => {
        //updated cached sorted sequences
        if (sortOption.isQuerySequenceDependent) {
          this.getSequences(sortOption, true);
        }
      });
      this.sequences.forEach(
        (sequences: ISequence[], sortBy: SequenceSorter) => {
          this.getSequences(sortBy, true); // force re-sorting, which updates cache
        }
      );
    } else {
      throw Error("query sequence does not exist in alignment");
    }
  }*/
}
