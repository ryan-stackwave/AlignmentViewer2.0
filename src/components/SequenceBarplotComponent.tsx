import React from "react";
import "./SequenceBarplot.scss";
import _ from "lodash";
import { Alignment } from "../common/Alignment";
import { mapGroupBy, ArrayOneOrMore } from "../common/Utils";
import ReactTooltip from "react-tooltip";

export interface ISequenceBarplotDataSeries {
  id: string; //must be unique for each series
  name: string;
  cssClass: string;
  plotOptions?: {
    fixYMax?(alignment: Alignment): number; //defaults to data max
    //fixYMin?: number; //defaults to data min
  };
  getPositionalInfo(pos: number, alignment: Alignment): ISingleBarDetails;
}

export interface ISequenceBarplotProps {
  alignment: Alignment;
  dataSeries: ArrayOneOrMore<ISequenceBarplotDataSeries>;
  positionWidth: number;

  height?: number;
  onPositionSelectionChanged?(
    positionsSelected: {
      position: number;
      element: SVGRectElement;
      bars: ISingleBarDetailsFull[];
    }[]
  ): void;
  onPositionMouseEnter?(
    position: number,
    element: SVGRectElement,
    bars: ISingleBarDetailsFull[]
  ): void;
  onPositionMouseLeave?(
    position: number,
    element: SVGRectElement,
    bars: ISingleBarDetailsFull[]
  ): void;
}

interface ISingleBarDetails {
  height: number;
}
interface ISingleBarDetailsFull extends ISingleBarDetails {
  normalizedHeight: number;
  position: number;
  dataSeries: ISequenceBarplotDataSeries;
}
interface ISequenceBarplotState {
  positionHovered?: {
    position: number;
    element: SVGRectElement;
    bars: ISingleBarDetailsFull[];
  };
  positionsSelected: {
    position: number;
    element: SVGRectElement;
    bars: ISingleBarDetailsFull[];
  }[];
  svgHovered: boolean;
}

export class SequenceBarplotComponent extends React.Component<
  ISequenceBarplotProps,
  ISequenceBarplotState
> {
  private static POSITION_VIEWBOX_HEIGHT = 100;
  private static POSITION_VIEWBOX_WIDTH = 1;

  private cache: Partial<ISequenceBarplotProps> & {
    bars?: ISingleBarDetailsFull[];
    barsGroupedByPosition?: Map<number, ISingleBarDetailsFull[]>;
    barsGroupedByDataseries?: Map<
      ISequenceBarplotDataSeries,
      ISingleBarDetailsFull[]
    >;
  } = {};

  static defaultProps = {
    height: 100,
  };

  constructor(props: ISequenceBarplotProps) {
    super(props);
    this.state = {
      svgHovered: false,
      positionsSelected: [],
    };
    this.handleSvgHover = this.handleSvgHover.bind(this);
    this.handlePositionHover = this.handlePositionHover.bind(this);
    this.handlePositionClicked = this.handlePositionClicked.bind(this);
  }

  /*
   *
   *
   * PRE-CONFIGURED BARPLOTS
   *
   *
   */
  /**
   * Plot the shannon entropy at each position -sum(p * log p)
   *   - p is the fraction of each upper case letter at the position
   *   - ymax is set to the maximum entropy, which is an equal number
   *     of all upper case letters in the alignment.
   */
  public static SHANNON_ENTROPY_BARPLOT: ISequenceBarplotDataSeries = {
    id: "entropy",
    name: "Entropy",
    cssClass: "barplot-shannon-entropy",
    plotOptions: {
      fixYMax: (al) => {
        const allLettersInAlignment = al.getAllUpperAlphaLettersInAlignmentSorted();
        const p = 1 / allLettersInAlignment.length;
        return (
          -1 *
          allLettersInAlignment.reduce((acc) => {
            return acc + p * Math.log2(p);
          }, 0)
        );
      },
    },
    getPositionalInfo: (pos, al) => {
      const plc = al
        .getPositionalLetterCounts(
          true,
          al.getAllUpperAlphaLettersInAlignmentSorted()
        )
        .get(pos);
      return {
        height:
          !plc || Object.keys(plc).length === 0
            ? 0
            : -1 *
              Object.values(plc).reduce((acc, p) => {
                return acc + p * Math.log2(p);
              }, 0),
      };
    },
  };

  /**
   * Plot the Kullback-Leibler (KL) divergence for individual columns in the alignment
   *        S = sum(pk * log(pk / qk)
   *   also see scipy docs:
   *     https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.entropy.html
   *
   * No maximum is explicitly set (so it because the max value)
   */
  public static KULLBAC_LEIBLER_DIVERGENCE_BARPLOT: ISequenceBarplotDataSeries = {
    id: "kullback-leibler-divergence",
    name: "KL Divergence",
    cssClass: "barplot-kullback-leibler-divergence",
    getPositionalInfo: (pos, al) => {
      const allLetters = al.getAllUpperAlphaLettersInAlignmentSorted();
      const pk = al.getPositionalLetterCounts(true, allLetters).get(pos);
      const qk = al.getGlobalAlphaLetterCounts(true, allLetters);
      return {
        height:
          !pk || Object.keys(pk).length === 0
            ? 0
            : allLetters.reduce((acc, letter) => {
                if (letter in pk) {
                  acc += pk[letter] * Math.log(pk[letter] / qk[letter]);
                }
                return acc;
              }, 0),
      };
    },
  };

  /**
   * Plot the number of gaps at each position. The y max is set to
   * the total length of the alignment and not the largest number of
   * gaps.
   */
  public static GAPS_BARPLOT: ISequenceBarplotDataSeries = {
    id: "gaps",
    name: "Gaps",
    cssClass: "barplot-gaps",
    plotOptions: {
      fixYMax: (alignment) => alignment.getSequences().length,
    },
    getPositionalInfo: (pos, al) => {
      return {
        height: al.getGapCountAtColumn(pos),
      };
    },
  };

  /**
   *
   *
   * HELPER FUNCTIONS
   *
   *
   */
  private handleSvgHover(isHovered: boolean) {
    this.setState({
      svgHovered: isHovered,
    });
  }

  private handlePositionHover(
    event: React.MouseEvent<SVGElement, MouseEvent>,
    position: number,
    isHovered: boolean
  ) {
    const { onPositionMouseEnter, onPositionMouseLeave } = this.props;
    if (isHovered && onPositionMouseEnter) {
      //notify parent listeners
      onPositionMouseEnter(
        position,
        event.relatedTarget as SVGRectElement,
        this.getBars().barsGroupedByPosition.get(position)!
      );
    } else if (!isHovered && onPositionMouseLeave) {
      onPositionMouseLeave(
        position,
        event.relatedTarget as SVGRectElement,
        this.getBars().barsGroupedByPosition.get(position)!
      );
    }
    this.setState({
      positionHovered: isHovered
        ? {
            position: position,
            element: event.relatedTarget as SVGRectElement,
            bars: this.getBars().barsGroupedByPosition.get(position)!,
          }
        : undefined,
    });
  }

  private handlePositionClicked(
    event: React.MouseEvent<SVGElement, MouseEvent>,
    position: number
  ) {
    const { positionsSelected } = this.state;
    const { onPositionSelectionChanged } = this.props;
    const newPositionsSelected = positionsSelected.some(
      (p) => p.position === position
    )
      ? positionsSelected.filter((p) => p.position !== position)
      : [
          ...positionsSelected,
          {
            position: position,
            element: event.relatedTarget as SVGRectElement,
            bars: this.getBars().barsGroupedByPosition.get(position)!,
          },
        ];
    if (onPositionSelectionChanged) {
      onPositionSelectionChanged(newPositionsSelected);
    }
    this.setState({
      positionsSelected: newPositionsSelected,
    });
  }

  /**
   * Normalize bars to fall between 0 and POSITION_MAX_HEIGHT in svg space. Obey the
   * fixYMax if specified
   * @param bars the same bars as input, with the "normalizeHeight"
   *             attribute added
   */
  private normalizeBarHeights(
    bars: ISingleBarDetailsFull[]
  ): ISingleBarDetailsFull[] {
    const { alignment } = this.props;
    const allHeights = bars.map((bar) => bar.height);
    const overallMinHeight = Math.min(...allHeights);
    const overallMaxHeight = Math.max(...allHeights);
    return bars.map((bar) => {
      const minHeight = overallMinHeight;
      const maxHeight =
        bar.dataSeries.plotOptions && bar.dataSeries.plotOptions.fixYMax
          ? bar.dataSeries.plotOptions.fixYMax(alignment)
          : overallMaxHeight;
      return {
        ...bar,
        normalizedHeight:
          ((bar.height - minHeight) / (maxHeight - minHeight)) *
          SequenceBarplotComponent.POSITION_VIEWBOX_HEIGHT,
      };
    });
  }

  /**
   * Get bars and bar groups.
   * Calls prop.getPositionalInfo to calculate heights for each
   * position. Uses cache to avoid too many updates.
   */
  private getBars() {
    const { alignment, dataSeries } = this.props;
    if (
      !this.cache.bars ||
      alignment !== this.cache.alignment ||
      dataSeries !== this.cache.dataSeries
    ) {
      this.cache.alignment = alignment;
      this.cache.dataSeries = dataSeries;

      const maxSeqLength = alignment.getMaxSequenceLength();
      let allBars: ISingleBarDetailsFull[] = [];
      dataSeries.forEach((ds) => {
        for (let i = 0; i < maxSeqLength; i++) {
          allBars.push({
            ...ds.getPositionalInfo(i, alignment),
            position: i,
            dataSeries: ds,
            normalizedHeight: -1, // define below
          });
        }
      });

      //normalize bars group by group
      this.cache.bars = Object.entries(
        _.groupBy(allBars, (bar) => bar.dataSeries.id)
      )
        .map(([categoryId, categoryBars]) => {
          return this.normalizeBarHeights(categoryBars);
        })
        .flat();

      this.cache.barsGroupedByPosition = mapGroupBy(
        this.cache.bars,
        (bar) => bar.position
      );

      this.cache.barsGroupedByDataseries = mapGroupBy(
        this.cache.bars,
        (item) => item.dataSeries
      );
    }
    return {
      bars: this.cache.bars!,
      barsGroupedByPosition: this.cache.barsGroupedByPosition!,
      barsGroupedByDataseries: this.cache.barsGroupedByDataseries!,
    };
  }

  /**
   *
   *
   * JSX RENDER FUNCTIONS
   *
   *
   */

  /**
   * Render the tooltip for a particular position.
   */
  private renderTooltip() {
    const barsObj = this.getBars();
    return (
      <ReactTooltip
        id="getBarTooltip"
        effect="solid"
        type="light"
        border={true}
        getContent={(pos: string) => {
          console.log("BAR TOOLTIP:" + pos);
          const posInt = parseInt(pos);
          const barsAtPostion = barsObj.barsGroupedByPosition.get(posInt)!;
          const posPlusOne = posInt + 1; // positions should be 1 based, not zero based
          return !barsAtPostion ? undefined : (
            <div className="bar-position-textblock">
              <h1>Position: {posPlusOne}</h1>
              {barsAtPostion.map((bar) => {
                return (
                  <div
                    className={`dataseries-line ${bar.dataSeries.cssClass}`}
                    key={bar.dataSeries.id}
                  >
                    <span className="legend-square"></span>
                    <span className="legend-text">
                      {bar.dataSeries.name}: {+bar.height.toFixed(1)}
                    </span>
                  </div>
                );
              })}
            </div>
          );
        }}
      />
    );
  }

  /**
   * Render the bar plot svg, with each bar appearing as a single rectangle.
   * Final form:
   *    <svg ...>
   *      <g position 1>
   *         <rect transparant rectangle for interaction for position 1>
   *         <rect for dataseries 1 position 1>
   *         <rect for dataseries 2 position 1>
   *      </g>
   *      <g position 2>
   *         <rect transparant rectangle for interaction for position 2>
   *         <rect for dataseries 1 position 2>
   *         <rect for dataseries 2 position 2>
   *      </g>
   *   </svg>
   */
  private renderBarPlot() {
    const { alignment, positionWidth, height } = this.props;

    const maxSeqLength = alignment.getMaxSequenceLength();
    const totalWidth = positionWidth * maxSeqLength;

    const barsObj = this.getBars();

    return (
      <svg
        preserveAspectRatio="none"
        viewBox={`0 0 ${
          maxSeqLength * SequenceBarplotComponent.POSITION_VIEWBOX_WIDTH
        } ${SequenceBarplotComponent.POSITION_VIEWBOX_HEIGHT}`}
        style={{ width: totalWidth, height: height }}
        xmlns="http://www.w3.org/2000/svg"
        onMouseEnter={() => this.handleSvgHover(true)}
        onMouseLeave={() => this.handleSvgHover(false)}
      >
        {Array.from(barsObj.barsGroupedByPosition.entries()).map(
          ([pos, bars]) => {
            const maxBarHeight = Math.max(
              ...bars.map((bar) => bar.normalizedHeight)
            );

            const numDataSeries = barsObj.barsGroupedByDataseries.size;
            const barWidth =
              (numDataSeries === 1 ? 0.9 : 1 / numDataSeries) *
              SequenceBarplotComponent.POSITION_VIEWBOX_WIDTH;
            const firstBarOffset =
              (numDataSeries === 1 ? pos + 0.05 : pos) *
              SequenceBarplotComponent.POSITION_VIEWBOX_WIDTH;

            return (
              <g
                transform={`translate(${firstBarOffset},0)`}
                className={
                  this.state.positionsSelected.some((p) => p.position === pos)
                    ? "position-container selected"
                    : "position-container"
                }
                onClick={(event) => {
                  this.handlePositionClicked(event, pos);
                }}
                onMouseEnter={(event) => {
                  this.handlePositionHover(event, pos, true);
                }}
                onMouseLeave={(event) => {
                  this.handlePositionHover(event, pos, false);
                }}
                data-for="getBarTooltip"
                data-tip={pos}
                data-class={"barplot-tooltip-container"}
                key={`pos_${pos}`}
              >
                {bars.map((bar, dataseriesIdx) => {
                  return (
                    <rect
                      className={bar.dataSeries.cssClass}
                      transform={`translate(${
                        ((dataseriesIdx * 1) / bars.length) *
                        SequenceBarplotComponent.POSITION_VIEWBOX_WIDTH
                      },${
                        SequenceBarplotComponent.POSITION_VIEWBOX_HEIGHT -
                        bar.normalizedHeight
                      })`}
                      width={barWidth}
                      height={bar.normalizedHeight}
                      key={`${bar.position}_${bar.dataSeries.id}`}
                    ></rect>
                  );
                })}

                <rect
                  className="interaction-placeholder"
                  transform={`translate(0,${
                    SequenceBarplotComponent.POSITION_VIEWBOX_HEIGHT -
                    maxBarHeight
                  })`}
                  width={SequenceBarplotComponent.POSITION_VIEWBOX_WIDTH}
                  height={maxBarHeight}
                ></rect>
              </g>
            );
          }
        )}
      </svg>
    );
  }

  /*
   *
   *
   * REACT METHODS
   *
   *
   */
  shouldComponentUpdate(nextProps: ISequenceBarplotProps) {
    const { alignment, dataSeries, positionWidth, height } = this.props;
    if (
      !this.cache ||
      alignment !== nextProps.alignment ||
      dataSeries !== nextProps.dataSeries ||
      positionWidth !== nextProps.positionWidth ||
      height !== nextProps.height
    ) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    ReactTooltip.rebuild();
  }

  render() {
    const className = this.state.svgHovered ? "barplot hovered" : "barplot";

    return !this.props.alignment ? null : (
      <div className={className}>
        {this.renderTooltip()}
        {this.renderBarPlot()}
      </div>
    );
  }
}
