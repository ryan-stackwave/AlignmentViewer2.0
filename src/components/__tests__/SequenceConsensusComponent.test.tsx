import "jest-webgl-canvas-mock";
import * as React from "react";

import { shallow } from "enzyme";
import { Alignment } from "../../common/Alignment";
import { SequenceBarplotComponent } from "../SequenceBarplotComponent";

describe("SequenceBarplotComponent", () => {
  it("Should render when given default props.", () => {
    const wrapper = shallow(
      <SequenceBarplotComponent
        positionWidth={3}
        alignment={new Alignment("", [])}
        dataSeries={[SequenceBarplotComponent.SHANNON_ENTROPY_BARPLOT]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
