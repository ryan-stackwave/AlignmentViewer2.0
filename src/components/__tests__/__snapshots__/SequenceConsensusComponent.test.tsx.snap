// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceBarplotComponent Should render when given default props. 1`] = `
<div
  className="barplot"
>
  <ReactTooltip
    border={true}
    clickable={false}
    effect="solid"
    getContent={[Function]}
    id="getBarTooltip"
    insecure={true}
    resizeHide={true}
    type="light"
    wrapper="div"
  />
  <svg
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    preserveAspectRatio="none"
    style={
      Object {
        "height": 100,
        "width": -3,
      }
    }
    viewBox="0 0 -1 100"
    xmlns="http://www.w3.org/2000/svg"
  />
</div>
`;
